File: R/pick.R
# Phase 15 -- true-geometry hit-testing.
#
# `scene_model()` reports bounding boxes. That is exactly right for a rectangular
# brush and wrong for anything diagonal or thin: a segment's bbox is the whole
# rectangle its endpoints span, so a bbox-distance "nearest" matches it from
# anywhere in that rectangle. A host with only boxes has to work around it --
# vellumwidget excludes graph edges from open-space hover for precisely this
# reason. These two functions supply the geometry the box was standing in for.

# The pick table for a scene at one probe point, in device px.
#
# Deliberately NOT a data frame: `key`/`kind`/`dist`/`n` are per element while
# `x`/`y` are per vertex, so coercing the whole thing would recycle the short
# columns and silently invent elements.
.pick_table <- function(scene, px, py) {
  scene <- as_vellum_scene(scene)
  .scene_to_backend(scene)$pick_table(px, py)
}

# npc/px -> device px, sharing `hit_test()`'s convention (npc y is up).
.probe_px <- function(scene, x, y, units) {
  d <- .scene_to_backend(as_vellum_scene(scene))$dim()
  if (identical(units, "npc")) c(x * d[1], (1 - y) * d[2]) else c(x, y)
}

#' Find the marks nearest a point, by true geometry
#'
#' Ranks a scene's keyed elements by their distance to a point — measuring to the
#' **actual geometry**, not to a bounding box.
#'
#' @section Why not the bounding box:
#' For a round or upright mark the two agree. For anything diagonal or thin they
#' do not, and the difference is not subtle: a line from the bottom-left of a
#' panel to the top-right has a bounding box covering the *entire panel*, so a
#' box-based "what is nearest" reports it from anywhere in the plot. That is why
#' a host holding only boxes has to exclude such marks from hover rather than
#' rank them wrongly.
#'
#' Distances are: to the disc for round marks (zero inside), to the rectangle for
#' rects (zero inside), perpendicular to the segment for segments, to the nearest
#' edge for open polylines, and **zero anywhere inside** a closed polygon or
#' path — so clicking the middle of a choropleth region hits that region.
#'
#' Marks vellum cannot describe more precisely than a box — labels, rounded rects
#' — fall back to their box, which for a label is a fair description of the
#' target anyway.
#'
#' @section For a client-side host:
#' A browser cannot call back into R on every mouse move. Use
#' [element_geometry()] instead: it returns the same true geometry once, and the
#' client computes distances locally. `vl_nearest()` is for hosts that *can* ask
#' — a Shiny app, a test, a script.
#'
#' @param scene A [vl_scene()] (or anything with an [as_vellum_scene()] method).
#' @param x,y The probe point.
#' @param units `"npc"` (default, y up) or `"px"` (device pixels, y down) — the
#'   same convention as [hit_test()].
#' @param n How many nearest elements to return.
#' @param max_dist Ignore anything further than this many device pixels.
#' @return A data frame of `key`, `kind` and `dist` (device px), nearest first.
#'   Zero rows if nothing is in range. Only *keyed* elements are considered —
#'   an unkeyed mark is not addressable, so reporting it would be no use.
#' @seealso [element_geometry()] for the geometry itself, [scene_model()] for
#'   bounding boxes, [hit_test()] for "which grob is under this pixel".
#' @examples
#' s <- vl_scene(4, 3, dpi = 96, bg = "white") |>
#'   draw(segments_grob(0.1, 0.1, 0.9, 0.9, key = "diagonal")) |>
#'   draw(points_grob(0.8, 0.2, key = "corner"))
#'
#' # Near the corner the diagonal's *bounding box* also contains this point,
#' # but its geometry is far away.
#' vl_nearest(s, 0.8, 0.2)
#' @export
vl_nearest <- function(scene, x, y, units = c("npc", "px"), n = 1L,
                       max_dist = Inf) {
  units <- match.arg(units)
  p <- .probe_px(scene, x, y, units)
  t <- .pick_table(scene, p[1], p[2])
  if (!length(t$key)) {
    return(data.frame(key = character(0), kind = character(0), dist = numeric(0)))
  }
  out <- data.frame(key = t$key, kind = t$kind, dist = t$dist,
                    stringsAsFactors = FALSE)
  out <- out[out$dist <= max_dist, , drop = FALSE]
  out <- out[order(out$dist), , drop = FALSE]
  utils::head(out, max(0L, as.integer(n)))
}

#' The true geometry of a scene's addressable elements
#'
#' Returns the resolved device-pixel geometry of every **keyed** element: the two
#' endpoints of a segment, the vertices of a line or polygon, the centre of a
#' round mark, the opposite corners of a rect.
#'
#' This is what a host needs to hit-test *exactly* without asking the engine per
#' event. [scene_model()] gives bounding boxes, which are right for a brush and
#' misleading for diagonal or thin marks; this gives the shape the box was
#' standing in for, once, so a browser can compute point-to-geometry distances
#' locally at whatever rate it likes.
#'
#' @section Reading the result:
#' One row per vertex, grouped by `key`. What the vertices mean depends on
#' `kind`:
#'
#' | `kind` | rows | meaning |
#' |---|---|---|
#' | `segment` | 2 | the endpoints |
#' | `line`, `polygon` | *k* | the vertices, in order; `polygon` is closed |
#' | `path` | *k* | all rings' vertices concatenated |
#' | `point` | 1 | the centre |
#' | `rect` | 2 | opposite corners |
#' | `text`, `roundrect` | 2 | the bounding box's corners |
#'
#' Coordinates are device pixels with y growing **down**, matching
#' [scene_model()]'s boxes and the rendered SVG's coordinate system.
#'
#' @inheritParams vl_nearest
#' @return A data frame of `key`, `kind`, `vertex` (1-based within the element),
#'   `x` and `y`. Zero rows if the scene has no keyed elements.
#' @seealso [vl_nearest()], [scene_model()]
#' @examples
#' s <- vl_scene(4, 3, dpi = 96, bg = "white") |>
#'   draw(segments_grob(0.1, 0.1, 0.9, 0.9, key = "diagonal"))
#' element_geometry(s)
#' @export
element_geometry <- function(scene) {
  # The probe point is irrelevant to the geometry; the walk computes both in one
  # pass, and (0, 0) keeps the distances it also returns cheap and meaningless
  # rather than subtly wrong-looking.
  t <- .pick_table(scene, 0, 0)
  empty <- data.frame(key = character(0), kind = character(0),
                      vertex = integer(0), x = numeric(0), y = numeric(0),
                      stringsAsFactors = FALSE)
  if (!length(t$key) || !length(t$x)) {
    return(empty)
  }
  n <- as.integer(t$n)
  data.frame(
    key = rep(t$key, n),
    kind = rep(t$kind, n),
    vertex = unlist(lapply(n, seq_len), use.names = FALSE),
    x = t$x, y = t$y,
    stringsAsFactors = FALSE
  )
}

File: src/rust/src/pick.rs
//! True-geometry hit-testing (Phase 15).
//!
//! `scene_model()` reports each element's *bounding box*, which is all a host
//! needs for a rectangular brush and is wrong for anything diagonal or thin. A
//! segment's bbox is the whole rectangle spanned by its endpoints, so a
//! bbox-distance "nearest" snaps to it from anywhere inside that rectangle —
//! which is why `vellumwidget` excludes graph edges from open-space hover
//! entirely (`srcts/index.ts:1316-1321`).
//!
//! This module supplies the geometry the box was standing in for.

/// Squared distance from a point to a segment.
pub fn seg_dist2(px: f64, py: f64, ax: f64, ay: f64, bx: f64, by: f64) -> f64 {
    let (dx, dy) = (bx - ax, by - ay);
    let len2 = dx * dx + dy * dy;
    let t = if len2 > 0.0 {
        (((px - ax) * dx + (py - ay) * dy) / len2).clamp(0.0, 1.0)
    } else {
        0.0
    };
    let (qx, qy) = (ax + t * dx, ay + t * dy);
    (px - qx).powi(2) + (py - qy).powi(2)
}

/// Is a point inside a ring? Even-odd crossing test.
pub fn point_in_ring(px: f64, py: f64, xs: &[f64], ys: &[f64]) -> bool {
    let n = xs.len().min(ys.len());
    if n < 3 {
        return false;
    }
    let mut inside = false;
    let mut j = n - 1;
    for i in 0..n {
        if (ys[i] > py) != (ys[j] > py) {
            let t = (py - ys[i]) / (ys[j] - ys[i]);
            if px < xs[i] + t * (xs[j] - xs[i]) {
                inside = !inside;
            }
        }
        j = i;
    }
    inside
}

/// Distance from a point to a polyline or ring, in the same units as the
/// coordinates.
///
/// `closed` treats the geometry as a ring: the last vertex joins the first, and
/// a point *inside* is at distance zero. That distinction is what makes a
/// filled polygon behave like a region rather than an outline — clicking the
/// middle of a country on a choropleth should hit it.
pub fn poly_dist(px: f64, py: f64, xs: &[f64], ys: &[f64], closed: bool) -> f64 {
    let n = xs.len().min(ys.len());
    if n == 0 {
        return f64::INFINITY;
    }
    if n == 1 {
        return ((px - xs[0]).powi(2) + (py - ys[0]).powi(2)).sqrt();
    }
    if closed && point_in_ring(px, py, xs, ys) {
        return 0.0;
    }
    let mut best = f64::INFINITY;
    for i in 0..n - 1 {
        best = best.min(seg_dist2(px, py, xs[i], ys[i], xs[i + 1], ys[i + 1]));
    }
    if closed {
        best = best.min(seg_dist2(px, py, xs[n - 1], ys[n - 1], xs[0], ys[0]));
    }
    best.sqrt()
}

/// Distance from a point to an axis-aligned rectangle; zero inside.
pub fn rect_dist(px: f64, py: f64, x0: f64, y0: f64, x1: f64, y1: f64) -> f64 {
    let (lo_x, hi_x) = (x0.min(x1), x0.max(x1));
    let (lo_y, hi_y) = (y0.min(y1), y0.max(y1));
    let dx = (lo_x - px).max(0.0).max(px - hi_x);
    let dy = (lo_y - py).max(0.0).max(py - hi_y);
    (dx * dx + dy * dy).sqrt()
}

/// Distance from a point to a circle's disc; zero inside.
pub fn circle_dist(px: f64, py: f64, cx: f64, cy: f64, r: f64) -> f64 {
    (((px - cx).powi(2) + (py - cy).powi(2)).sqrt() - r.abs()).max(0.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn segment_distance_is_perpendicular_not_bbox() {
        // The case that motivates the module: a point near the *corner* of a
        // diagonal's bbox is far from the diagonal itself. A bbox test would
        // report zero.
        let d = poly_dist(0.0, 10.0, &[0.0, 10.0], &[0.0, 10.0], false);
        assert!((d - (50.0f64).sqrt()).abs() < 1e-9, "got {d}");
        assert_eq!(rect_dist(0.0, 10.0, 0.0, 0.0, 10.0, 10.0), 0.0, "bbox says inside");
    }

    #[test]
    fn a_point_on_the_segment_is_at_zero() {
        assert!(poly_dist(5.0, 5.0, &[0.0, 10.0], &[0.0, 10.0], false) < 1e-12);
    }

    #[test]
    fn distance_clamps_to_the_endpoints() {
        // Beyond the end, the nearest point is the endpoint, not the infinite line.
        let d = poly_dist(-3.0, 0.0, &[0.0, 10.0], &[0.0, 0.0], false);
        assert!((d - 3.0).abs() < 1e-9, "got {d}");
    }

    #[test]
    fn a_filled_ring_contains_its_interior() {
        let xs = [0.0, 10.0, 10.0, 0.0];
        let ys = [0.0, 0.0, 10.0, 10.0];
        assert_eq!(poly_dist(5.0, 5.0, &xs, &ys, true), 0.0);
        // Open, the same point is 5 from the nearest edge.
        assert!((poly_dist(5.0, 5.0, &xs, &ys, false) - 5.0).abs() < 1e-9);
        // Outside is the same either way.
        assert!((poly_dist(15.0, 5.0, &xs, &ys, true) - 5.0).abs() < 1e-9);
    }

    #[test]
    fn point_in_ring_handles_a_concave_shape() {
        // An L: the notch is outside even though it is inside the bbox.
        let xs = [0.0, 10.0, 10.0, 4.0, 4.0, 0.0];
        let ys = [0.0, 0.0, 4.0, 4.0, 10.0, 10.0];
        assert!(point_in_ring(2.0, 2.0, &xs, &ys));
        assert!(!point_in_ring(8.0, 8.0, &xs, &ys), "the notch is outside");
    }

    #[test]
    fn circle_distance_is_to_the_disc() {
        assert_eq!(circle_dist(0.0, 0.0, 0.0, 0.0, 5.0), 0.0);
        assert!((circle_dist(10.0, 0.0, 0.0, 0.0, 5.0) - 5.0).abs() < 1e-9);
    }

    #[test]
    fn rect_distance_is_zero_inside_and_euclidean_outside() {
        assert_eq!(rect_dist(5.0, 5.0, 0.0, 0.0, 10.0, 10.0), 0.0);
        assert!((rect_dist(13.0, 14.0, 0.0, 0.0, 10.0, 10.0) - 5.0).abs() < 1e-9);
    }

    #[test]
    fn degenerate_geometry_does_not_panic() {
        assert!(poly_dist(0.0, 0.0, &[], &[], false).is_infinite());
        assert_eq!(poly_dist(3.0, 4.0, &[0.0], &[0.0], false), 5.0);
        assert!(!point_in_ring(0.0, 0.0, &[0.0, 1.0], &[0.0, 1.0]));
    }
}
